### PR TITLE
setup.py: add dependencies

### PR DESCRIPTION
**Pull request by kgugala**
This PR adds python dependencies to setup.py. In the binary flow we install this using pip and we were missing the deps (accidentally we had them installed by some other packages).

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -24,6 +24,25 @@
     long_description_content_type="text/markdown",
     url="https://github.com/SymbiFlow/prjxray",
     packages=['prjxray'],
+    install_requires=[
+        'sdf-timing @ git+https://github.com/symbiflow/python-sdf-timing',
+        'fasm @ git+https://github.com/symbiflow/fasm',
```
Comment (litghost) [inline setup.py:29]: Don't use an `@` in `setup.py`.  If you need to pull in a git package, do it via `requirements.txt`.

File: setup.py [hunk 2]
```diff
@@ -24,6 +24,25 @@
...
+        'parse',
+        'progressbar2',
+        'pyjson5',
+        'pytest',
+        'pyyaml',
+        'scipy>=1.2.1',
+        'simplejson',
+        'sympy',
+        'textx',
+        'yapf==0.24.0',
```
Comment (litghost) [inline setup.py:44]: `yapf` isn't need for end users?

File: setup.py [hunk 3]
```diff
@@ -24,6 +24,25 @@
     long_description_content_type="text/markdown",
     url="https://github.com/SymbiFlow/prjxray",
     packages=['prjxray'],
+    install_requires=[
+        'sdf-timing @ git+https://github.com/symbiflow/python-sdf-timing',
+        'fasm @ git+https://github.com/symbiflow/fasm',
+        'intervaltree',
+        'junit-xml',
```
Comment (litghost) [inline setup.py:31]: junit-xml is definitely not needed in setup.py?

File: setup.py (hunk 1 above)
Comment (kgugala) [reply setup.py:29]: the problem this PR solves is that those were not installed when we do `pip install git+https://github.com/symbiflow/prjxray` (and we do that in symbiflow-examples). That is why I put those `@` in `setup.py`

File: setup.py (hunk 2 above)
Comment (kgugala) [reply setup.py]: removed

File: setup.py (hunk 1 above)
Comment (litghost) [reply setup.py:29]: Never put `@` in setup.py, end of story.  It causes all kinds of problems.  `@` is an extremely narrow pin, and causes problems.  The longer term solution is for these packages to end up on PyPi, etc, but that is a longer term problem.

File: setup.py (hunk 3 above)
Comment (kgugala) [reply setup.py]: removed

File: setup.py (hunk 1 above)
Comment (ajelinski) [reply setup.py:29]: > `@` is an extremely narrow pin

Can you explain why @litghost? Does it tie a dependency like `fasm` to the revision which is currently `master` on `fasm` forever if I install a package with `fasm @ git+https://github.com/symbiflow/fasm` in `install_requires`?

Comment (litghost) [reply setup.py:29]: The `@` is effectively "always use exactly this dependency".  The new pip dependency resolver respects this directly, but it means that it mixes poorly with other configurations.  As a concrete example, if you want to use `git submodule` to pin fasm to a particular commit (for example in symbiflow-examples), that will conflict with the `@` here.

The bottom line is `@` can **never** be used in `setup.py`, period.  There are no acceptable reasons to do so.

**Comment by ajelinski** [comment]
Definitely `pyyaml` is needed in `install_required`. If it was included, I wouldn't need to add it manually to `requirements.txt` files in https://github.com/SymbiFlow/fpga-tool-perf .

**Comment by mithro** [comment]
> Definitely `pyyaml` is needed in `install_required`. If it was included, I wouldn't need to add it manually to `requirements.txt` files in https://github.com/SymbiFlow/fpga-tool-perf .

Specifying `pyyaml` in the `install_requires` is very different to specifying `pyyaml @ <location>`.

File: setup.py [hunk 4]
```diff
@@ -24,6 +24,23 @@
     long_description_content_type="text/markdown",
     url="https://github.com/SymbiFlow/prjxray",
     packages=['prjxray'],
+    install_requires=[
+        'sdf-timing @ git+https://github.com/symbiflow/python-sdf-timing',
```
Comment (mithro) [inline setup.py:28]:
```suggestion
        'sdf-timing',
```

File: setup.py [hunk 5]
```diff
@@ -24,6 +24,23 @@
...
+        'fasm @ git+https://github.com/symbiflow/fasm',
```
Comment (mithro) [inline setup.py:29]:
```suggestion
        'fasm',
```

File: setup.py [hunk 6]
```diff
@@ -24,6 +24,23 @@
...
+        'ordered-set',
+        'parse',
+        'progressbar2',
+        'pyjson5',
+        'pytest',
+        'pyyaml',
+        'scipy>=1.2.1',
+        'simplejson',
+        'sympy',
+        'textx',
```
Comment (mithro) [inline setup.py:42]: Are all these dependencies *actually* needed? These seem like the dependencies for the whole prjxray fuzzer tooling not the prjxray python library?

File: setup.py (hunk 1 above)
Comment (litghost) [reply setup.py:29]: `fasm` is now on PyPi (per https://github.com/SymbiFlow/fasm/pull/59).  If someone can do the same work for sdf-timing, the `@` can go away from this completely!

**Comment by mithro** [comment]
The FASM library should now be installable via both PyPi and directly from GitHub.

The `install_requires` in `setup.py` should **only** specify `fasm`.

The `requirements.txt` can specify *any* of `fasm` (which will be downloaded via PyPi) or something like `git+https://github.com/SymbiFlow/fasm.git#egg=fasm` (downloaded and installed from GitHub) or `-e ./third_party/fasm` (installed from local third party submodule).

File: setup.py (hunk 6 above)
Comment (litghost) [reply setup.py:42]: Poke, this comment is valid, and should be fixed.

Comment (acomodi) [reply setup.py]: Should be fixed now. The only dependencies added are ones that are needed by scripts in the prjxray directory scripts

File: setup.py [hunk 7]
```diff
@@ -24,6 +24,14 @@
     long_description_content_type="text/markdown",
     url="https://github.com/SymbiFlow/prjxray",
     packages=['prjxray'],
+    install_requires=[
+        'fasm',
+        'intervaltree',
+        'numpy',
+        'pyjson5',
```
Comment (litghost) [inline setup.py:31]: Where is `pyjson5` used?

Comment (litghost) [reply setup.py:31]: And what happened to `sdf-timing`?  Is it not required?

Comment (acomodi) [reply setup.py:31]: > Where is pyjson5 used?

https://github.com/SymbiFlow/prjxray/blob/master/prjxray/lib.py
https://github.com/SymbiFlow/prjxray/blob/master/prjxray/node_lookup.py

> And what happened to sdf-timing

This seems not to be used by any of the scripts belonging to the python package

Comment (litghost) [reply setup.py:31]: > > Where is pyjson5 used?
> 
> https://github.com/SymbiFlow/prjxray/blob/master/prjxray/lib.py
> https://github.com/SymbiFlow/prjxray/blob/master/prjxray/node_lookup.py

Both of these probably shouldn't be in the prjxray library, as they operate on data that is only present in prjxray fuzzer directories.   Let's leave pyjson5 in for now, but add a README to isolate those files and eventually remove this dependency.

File: setup.py [hunk 8]
```diff
@@ -24,6 +24,15 @@
     long_description_content_type="text/markdown",
     url="https://github.com/SymbiFlow/prjxray",
     packages=['prjxray'],
+    install_requires=[
+        'fasm',
+        'intervaltree',
+        'numpy',
+        # FIXME: isolate lib.py and node_lookup.py from the package to remove this dependency
```
Comment (litghost) [inline setup.py:31]: It's specifically `lib.NodeLookup`, not all of `lib.py` that is a problem.  Also please fill an issue to track fixing this, and link it here.

**Review by litghost — APPROVED** [review]
LGTM, merge if green

**Pull request merged**